### PR TITLE
[GCS]Use an asynchronous PING to avoid blocking other operations

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
…erations

## Why are these changes needed?

We use same redis client same with other components, which takes main io service as execution thread, to detect redis failure using a synchronous command `PING`.
The check command is scheduled periodically with a small time value and it will block other operations in main thread easily as it keeps thread wait util the command returns. The blocking may slow down requests handling visiably.
We can init another redis client to detect redis failure using a separate io service and a thread.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Review by rkooo567 — COMMENTED** [review]
I am actually not sure if this change is necessary. It seems like we only have 10 calls to Redis per second, and we won't have more checks than that (meaning failure check is not linearly increasing to the cluster sizes). I don't think 10 asynchronous calls per seconds will affect the main workload (if so, that sounds like a huge problem from GCS).

File: src/ray/core_worker/lib/java/io_ray_runtime_task_NativeTaskSubmitter.cc [hunk 1]
```diff
@@ -197,9 +197,11 @@ JNIEXPORT jobject JNICALL Java_io_ray_runtime_task_NativeTaskSubmitter_nativeSub
...
   std::vector<ObjectID> return_ids;
   // TODO (kfstorm): Allow setting `max_retries` via `CallOptions`.
-  ray::CoreWorkerProcess::GetCoreWorker().SubmitTask(ray_function, task_args,
-                                                     task_options, &return_ids,
-                                                     /*max_retries=*/0);
+  ray::CoreWorkerProcess::GetCoreWorker().SubmitTask(
+      ray_function, task_args, task_options, &return_ids,
+      /*max_retries=*/0,
+      /*placement_options=*/
+      std::pair<ray::PlacementGroupID, int64_t>(ray::PlacementGroupID::Nil(), 0));
```
[the remark this answers is not in the document]
Comment (WangTaoTheTonic) [reply src/ray/core_worker/lib/java/io_ray_runtime_task_NativeTaskSubmitter.cc:204]: yes, it's unrelated but just to fix the master branch build. see https://github.com/ray-project/ray/pull/9868

**Comment by WangTaoTheTonic** [comment]
> I am actually not sure if this change is necessary. It seems like we only have 10 calls to Redis per second, and we won't have more checks than that (meaning failure check is not linearly increasing to the cluster sizes). I don't think 10 asynchronous calls per seconds will affect the main workload (if so, that sounds like a huge problem from GCS).

No, it is not asynchronous but SYNChronous PING.

Even in most time the latency is very small (like ~1ms), but sometimes the latency can be affected by network load and redis pressure, which will get higher to 5ms even more(30-60ms occasionally). That means every 100ms gcs would wait this PING command for ~5ms frequently (more than 10ms occasionally), which brings a >5% performance down.

There are lots functions using io service in same thread, which impact each other. We need to pick up those who blocks (like this one) and who will be easily affected by others(like node failure detector).

**Comment by rkooo567** [comment]
@WangTaoTheTonic I see. It'll be great if you can add that information to the description. I totally got the point of this PR now. I have another question in that case. Is there any way to just make PING asynchronous? I think that's the ideal solution in this case (than adding an extra thread).

**Comment by WangTaoTheTonic** [comment]
> @WangTaoTheTonic I see. It'll be great if you can add that information to the description! I totally got the point of this PR now. I have another question in that case. Is there any way to just make PING asynchronous? I think that's the ideal solution in this case (than adding an extra thread).

That's fair to make PING asynchronous.
The change has been applied. (Spent a long day with struggling with issue caused by [null pointer](https://github.com/ray-project/ray/pull/9871/files#diff-296be4a087fb421914ac7ab3f80b1508R100). )

**Review by rkooo567 — APPROVED** [review]
Thanks!! LGTM! And yeah nullptr is evil ;)

**Pull request merged**